File: src/api/middleware/NativeClientInfoMiddleware.ts
import RequestContext from "../model/RequestContext";
import ErrorCode from "../enum/ErrorCode";
import ApiError from "../error/ApiError";

export default async function (req, res) {
  const versionRegex = /^(\d+\.)?(\d+\.)?(\*|\d+)$/;

  const osType = <string>req.headers['x-os-type'];
  const osVersion = <string>req.headers['x-os-version'];
  const appType = <string>req.headers['x-app-type'];
  const appVersion = <string>req.headers['x-app-version'];

  const context: RequestContext = req['context'];

  if (osType != undefined) {
    if (osType === 'ios' || osType === 'android') {
      context.osType = osType;
    } else {
      throw new ApiError(400, ErrorCode.INVALID_OS_TYPE, `Invalid os type: ${osType}`);
    }
  }

  if (osVersion != undefined) {
    if (osVersion.match(versionRegex)) {
      context.osVersion = osVersion;
    } else {
      throw new ApiError(400, ErrorCode.INVALID_OS_VERSION, `Invalid os version: ${osVersion}`);
    }
  }

  if (appType != undefined) {
    if (appType === 'release' || appType === 'debug') {
      context.appType = appType;
    } else {
      throw new ApiError(400, ErrorCode.INVALID_APP_TYPE, `Invalid app type: ${appType}`);
    }
  } else {
    context.appType = 'release';
  }

  if (appVersion != undefined) {
    if (osType === 'android') {
      if (!appVersion.split('-')[0].match(versionRegex)) {
        throw new ApiError(400, ErrorCode.INVALID_APP_VERSION, `Invalid app version: ${appVersion}`);
      }
    }
    if (!appVersion.match(versionRegex)) {
      throw new ApiError(400, ErrorCode.INVALID_APP_VERSION, `Invalid app version: ${appVersion}`);
    }
    context.appVersion = appVersion;
  }

  return Promise.resolve('next');
}

File: api/src/main/kotlin/com/wafflestudio/snu4t/middleware/NativeClientInfoMiddleware.kt
package com.wafflestudio.snu4t.middleware

import com.wafflestudio.snu4t.common.client.AppType
import com.wafflestudio.snu4t.common.client.ClientInfo
import com.wafflestudio.snu4t.common.client.OsType
import org.springframework.stereotype.Component

@Component
class NativeClientInfoMiddleware() : BaseMiddleware({ req, context ->
    // TODO: 버전 검증 로직 필요
    // val versionRegex = """^(\d+\.)?(\d+\.)?(\*|\d+)$""".toRegex()

    // TODO: 오류 정의해야 함
    context.copy(
        clientInfo = ClientInfo(
            osType = OsType.fromString(req.headers().firstHeader("x-os-type")),
            osVersion = req.headers().firstHeader("x-os-version"),
            appType = AppType.fromString(req.headers().firstHeader("x-app-type")) ?: AppType.RELEASE,
            appVersion = req.headers().firstHeader("x-app-version"),
        )
    )
})
